File: 21/main.py
#######################################################
#####	21. Merge two sorted linked lists	#######
#######################################################


'''
The approach is the same. create a new list and apply the same merge logic from merge sort
Time Complexity:O(n+m) where n and m are the length of lists
Space Complexity:O(m+n) as we create a new list containing all the elements from input
'''

class Solution:
    def mergeTwoLists(self, l1, l2):
        dummy = cur = ListNode(0)
        while l1 and l2:
            if l1.val < l2.val:
                cur.next = l1
                l1 = l1.next
            else:
                cur.next = l2
                l2 = l2.next
            cur = cur.next
        cur.next = l1 or l2
        return dummy.next

File: 21/main.java
/**
*
*	13. Merge two sorted linked lists
*
**/


/*
We need to perform same logic as the merge logic of merge sort
Time Complexity:O(n) because we traverse each element of both lists once
Space Complexity:O(n) as we store each input element
*/

class Solution {
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        ListNode prev= new ListNode(0);
        ListNode ans=prev;
        
            while(l1!=null || l2!=null){
                if (l1==null){
                    ListNode cur=new ListNode(l2.val);
                    prev.next=cur;
                    prev=cur;
                    l2=l2.next;
                }
                else if (l2==null){
                    ListNode cur=new ListNode(l1.val);
                    prev.next=cur;
                    prev=cur;
                    l1=l1.next; 
                }
                else if(l1.val<=l2.val){
                    ListNode cur=new ListNode(l1.val);
                    prev.next=cur;
                    prev=cur;
                    l1=l1.next; 
                    
                    
                }
                else{
                    ListNode cur=new ListNode(l2.val);
                    prev.next=cur;
                    prev=cur;
                    l2=l2.next; 

                }
                
                
            }
        return ans.next;
        
        
    }
}
